hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/upgrade/TwoToOneDowngradeHandler.java

```diff
@@ -0,0 +1,121 @@
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Downgrade handle to assist in downgrading hoodie table from version 2 to 1.
+ */
+public class TwoToOneDowngradeHandler implements DowngradeHandler {
+
+  @Override
+  public void downgrade(HoodieWriteConfig config, HoodieEngineContext context, String instantTime) {
+    HoodieSparkTable table = HoodieSparkTable.create(config, context);
+    HoodieTableMetaClient metaClient = table.getMetaClient();
+    Properties properties = metaClient.getTableConfig().getProps();
+    properties.remove(HoodieTableConfig.HOODIE_TABLE_PARTITION_FIELDS_PROP.key());
+    properties.remove(HoodieTableConfig.HOODIE_TABLE_RECORDKEY_FIELDS.key());
+    properties.remove(HoodieTableConfig.HOODIE_BASE_FILE_FORMAT_PROP.key());
+
+    // Serializes the updated properties
+    // Since the metaclient object is not shared between AbstractUpgradeDownGrade,
+    // we have to serialize the properties here
+    Path propertyFile = new Path(metaClient.getMetaPath() + "/" + HoodieTableConfig.HOODIE_PROPERTIES_FILE);
+    try (FSDataOutputStream os = metaClient.getFs().create(propertyFile)) {
+      properties.store(os, "");
+    } catch (IOException e) {
+      throw new HoodieIOException("Updating hoodie.properties file to remove some of the new props failed ", e);
+    }
+
+    // re-create marker files if any partial timeline server based markers are found
+    HoodieTimeline inflightTimeline = metaClient.getCommitsTimeline().filterPendingExcludingCompaction();
+    List<HoodieInstant> commits = inflightTimeline.getReverseOrderedInstants().collect(Collectors.toList());
+    for (HoodieInstant commitInstant : commits) {
+      // Converts the markers in new format to old format of direct markers
+      convertToDirectMarkers(
+          commitInstant.getTimestamp(), table, context, config.getMarkersDeleteParallelism());
+    }
+  }
+
+  /**
+   * Converts the markers in new format(timeline server based) to old format of direct markers,
+   * i.e., one marker file per data file, without MARKERS.type file.
+   *
+   * @param commitInstantTime instant of interest for marker conversion.
+   * @param table             instance of {@link HoodieSparkTable} to use
+   * @param context           instance of {@link HoodieEngineContext} to use
+   * @param parallelism       parallelism to use
+   */
+  private void convertToDirectMarkers(final String commitInstantTime,
+                                      HoodieSparkTable table,
+                                      HoodieEngineContext context,
+                                      int parallelism) {
+    String markerDir = table.getMetaClient().getMarkerFolderPath(commitInstantTime);
+    FileSystem fileSystem = FSUtils.getFs(markerDir, context.getHadoopConf().newCopy());
+    Option<String> markerTypeOption = MarkerUtils.readMarkerType(fileSystem, markerDir);
+    if (markerTypeOption.isPresent()) {
```

Comment:
There is some work here to ensure idempotency incase of failures. lets sync up sometime. I can go over the details.